sw/device/silicon_creator/lib/drivers/keymgr.c

```diff
@@ -0,0 +1,95 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/keymgr.h"
+
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"  // Generated.
+
+static const uint32_t base = TOP_EARLGREY_KEYMGR_BASE_ADDR;
+
+/**
+ * Checks the key manager `expected_state`.
+ *
+ * This function reads and clears the status and error code registers.
+ *
+ * @return `kErrorOk` if the key manager is at the `expected_state` and the
+ * status is idle or success.
+ */
+static rom_error_t check_expected_state(uint32_t expected_state) {
```

Comment:
Shouldn't this wait until keymgr is idle? IIRC, state transitions involve KMAC operations that take some time to complete.

Reply:
I could add an idle check, but I think in practice my assumption is that the key manager will advance faster than Ibex and the bus. Specially if we have other operations in between.

Reply:
Can we be certain of this?  Which clock domain is the keymgr in and how many clocks does it take to advance states?  Considering the diversity of clock domain specifications in `//sw/device/lib/arch/device_*.c`, I think we need to back this assumption up with some data.

Reply:
I have added an `WIP` polling check to account for cases in which the keymgr may be stuck processing.  I will document the advance state and register access times to be able to answer this question with data points.